mm2src/coins/eth/eth_withdraw.rs

```diff
@@ -0,0 +1,352 @@
+            wei_from_big_decimal, EthCoinType, EthPrivKeyPolicy, WithdrawError, WithdrawRequest, WithdrawResult,
+            ERC20_CONTRACT};
+use crate::eth::{Action, EthTxFeeDetails, KeyPair, SignedEthTx, UnSignedEthTx};
+use crate::rpc_command::init_withdraw::{WithdrawInProgressStatus, WithdrawTaskHandle};
+use crate::{BytesJson, EthCoin, TransactionDetails};
+use async_trait::async_trait;
+use bip32::DerivationPath;
+use common::custom_futures::timeout::FutureTimerExt;
+use common::now_sec;
+use crypto::{CryptoCtx, HwRpcError};
+use ethabi::Token;
+use ethkey::public_to_address;
+use futures::compat::Future01CompatExt;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::{MapToMmResult, MmError, OrMmError};
+use std::ops::Deref;
+
+cfg_wasm32! {
+    use web3::types::TransactionRequest;
+}
+
+#[async_trait]
+pub trait EthWithdraw
+where
+    Self: Sized + Sync,
+{
+    fn coin(&self) -> &EthCoin;
+
+    fn request(&self) -> &WithdrawRequest;
+
+    #[allow(clippy::result_large_err)]
+    fn on_generating_transaction(&self) -> Result<(), MmError<WithdrawError>>;
+
+    #[allow(clippy::result_large_err)]
+    fn on_finishing(&self) -> Result<(), MmError<WithdrawError>>;
+
+    async fn sign_tx_with_trezor(
+        &self,
+        derivation_path: DerivationPath,
+        unsigned_tx: &UnSignedEthTx,
+    ) -> Result<SignedEthTx, MmError<WithdrawError>>;
+
+    async fn build(self) -> WithdrawResult {
+        let coin = self.coin();
+        let ticker = coin.deref().ticker.clone();
+        let req = self.request().clone();
+
+        let to_addr = coin
+            .address_from_str(&req.to)
+            .map_to_mm(WithdrawError::InvalidAddress)?;
+        let (my_balance, my_address, key_pair, derivation_path) = match req.from {
+            Some(from) => {
+                let path_to_coin = &coin
+                    .deref()
+                    .derivation_method
+                    .hd_wallet()
+                    .ok_or(WithdrawError::UnexpectedDerivationMethod)?
+                    .derivation_path;
+                let path_to_address = from.to_address_path(path_to_coin.coin_type())?;
+                let derivation_path = path_to_address.to_derivation_path(path_to_coin)?;
+                let (key_pair, address) = match coin.priv_key_policy {
+                    EthPrivKeyPolicy::Trezor {
+                        ref activated_pubkey, ..
+                    } => {
+                        let my_pubkey = activated_pubkey
+                            .as_ref()
+                            .or_mm_err(|| WithdrawError::InternalError("empty trezor xpub".to_string()))?;
+                        let my_pubkey = pubkey_from_xpub_str(my_pubkey)
+                            .map_to_mm(|_| WithdrawError::InternalError("invalid trezor xpub".to_string()))?;
+                        let address = public_to_address(&my_pubkey);
+                        (None, address)
```

Comment:
We should be able to withdraw from any eth address for trezor similar to HD wallet and UTXO trezor impl, right now, we can only withdraw from the address enabled on coin activation.

Reply:
ooh, I see that you use the derivation path when signing the transaction https://github.com/KomodoPlatform/komodo-defi-framework/blob/958617caeac3967729bc1e0e084e904ebb4eef77/mm2src/coins/eth/eth_withdraw.rs#L175 but balance calculations use the address from the `activated_pubkey` which is not right.

Reply:
Yes, I see thank you, will fix the trezor case this for derivation_path (this part was not finished, I guess)

Reply:
fixed this bug in 4988565
also allowed to withdraw eth with trezor if 'from' is not set

Reply:
I can see more issues in this code...

Reply:
Changed the fix for getting 'from' hd account ([1569184](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2005/commits/1569184e77887bc7fcb5b6bfeb209558feb2d1d5)): now I use the cached account from the hd wallet instead of getting it from the device to avoid disruption due to a pin request. 
(Basically we should not call the device from internal code but use cached data. The device should be called from explicit rpc calls which can handle device callback like pin requests)

Reply:
Plz note I changed get_enabled_address() to get also derivation_path from it